libs/knowledge-store/ragstack_knowledge_store/document_graph.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+import itertools
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Optional, Union, Iterator, Tuple
+
+from langchain_core.callbacks import CallbackManagerForRetrieverRun
+from langchain_core.documents import Document
+from langchain_core.retrievers import BaseRetriever
+from pydantic import Field
+
+
+def _texts_to_documents(
+    texts: Iterator[str],
+    metadatas: Iterator[dict],
+) -> Iterator[Document]:
+    for (
+        text,
+        metadata,
+    ) in zip(texts, metadatas):
+        yield Document(
+            page_content=text,
+            metadata=metadata,
+        )
+
+
+class DocumentGraph(ABC):
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[Iterable[dict]] = None,
+        *,
+        ids: Optional[Iterable[str]] = None,
+        links: Iterable[Tuple[str, str]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        metadatas_it = iter(metadatas) if metadatas else itertools.repeat({})
+        documents = _texts_to_documents(iter(texts), metadatas_it)
+        return self.add_documents(documents, ids=ids, links=links, **kwargs)
+
+    @abstractmethod
+    def add_documents(
+        self,
+        documents: Iterable[Document] = None,
+        *,
+        ids: Optional[Iterable[str]] = None,
+        links: Iterable[Tuple[str, str]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        """Add documents to the graph"""
+
+    @abstractmethod
+    def traversing_retrieve(
+        self,
+        query: Union[str, Iterable[str]],
+        *,
+        k: int = 4,
+        depth: int = 1,
+        **kwargs: Any,
+    ) -> Iterable[Document]:
+        """Retrieve documents from traversing this knowledge store.
+
+        First, `k` nodes are retrieved using a search for each `query` string.
+        Then, additional nodes are discovered up to the given `depth` from those
+        starting nodes.
+
+        Args:
+            query: The query string or collection of query strings.
+            k: The number of Documents to return from the initial search.
+                Defaults to 4. Applies to each of the query strings.
+            depth: The maximum depth of edges to traverse. Defaults to 1.
+        Returns:
+            Retrieved documents.
+        """
+
+    def as_retriever(
+        self, retrieval_kwargs: dict[str, Any], **kwargs: Any
+    ) -> "DocumentGraphRetriever":
+        """Return a Retriever for retrieving from this knowledge store..
+
+        The initial nodes are retrieved using a search.
+        Additional nodes are discovered up to the given depth from those starting nodes.
+
+        Args:
+            retrieval_kwargs: Keyword arguments to pass to the retrieval function.
```

Comment:
I don't love this pattern -- but maybe with the `mypy` overload we can at least document the "standard" set of arguments?

Reply:
This is the pattern used for the VectorStoreRetriever.
It's done like this since the search function and its attributes can vary base on the search_type param.

Reply:
That makes sense. We can do the `@overload` to convey what the attributes are for a specific search type.

(although, we could call this `as_retriever_traverse` or `as_traversing_retriever`?

But, if we want `search_type`:

```
@overload
def as_retriever(self, search_type: Literal["traverse"], *, k: int = 4, depth: int = 1, **kwargs): ...

@overload
def as_retriever(self, search_type: Literal["mmr"], *, n: int = 10, max_depth: int = 2, **kwargs): ...

def as_retriever(self, search_type: str, **kwargs):
   ...
```

This let's us have types (and maybe docs) for specific search types, will still preserving the pattern from `VectorStoreRetriever` of allowing attributes to vary based on search_type.

Reply:
For now, removing `retrieval_kwargs`.

Reply:
Done. `as_retriever` now returns a Retriever that supports `traversal` as a search type.